release_notes/topics/19_0_0.adoc

```diff
@@ -25,3 +25,11 @@ Spring Boot 1.x reached end of life in 2019. If you are still using Spring Boot
 In WildFly 25 the legacy security layer was removed, going forward only Elytron will be supported. We recommend anyone using an older version of WildFly to upgrade and leverage native OIDC support in WildFly.
 
 Red Hat customers using Red Hat JBoss Enterprise Application Platform 7.x should use https://access.redhat.com/products/red-hat-single-sign-on[Red Hat Single Sign-On] 7.x adapters. These can be used in combination with the Keycloak server.
+
+= Handling of local vs. federated storage
+
+In preparation for the Storage.X map storage functionality, APIs have been consolidated so that the service layer won't differentiate between cached vs. non-cached, and federated vs. local storage.
+Also, the APIs for managing the credentials of users via the Java APIs from within {project_name} have moved to `UserModel.credentialManager()`.
+
+Those who wrote {project_name} extensions that involve the storage APIs should review the migration guide to find out about the changed semantics in the deprecated methods especially on `KeycloakSession`.
+This also describes how to access the old functionality using the legacy modules while those are available and before their code has been migrated to the new map storage APIs.
```

Comment:
```suggestion
= Changes in Keycloak storage

The Keycloak storage is changing and the current storage, while still supported, will eventually
be replaced with a brand new implementation. This change brings better support for cloud-native
storages, no-downtime abilities, and better support for implementing custom storages
for additional areas apart from users.

It means several deep changes in the supported features of current store which becomes _legacy_.
The legacy store and the new store cannot be used simultaneously - only one of them can be active
at the same time.

Most visibly, the user storage SPI is incompatible with the new storage (called map storage) API.
Thus User storage SPI will be deprecated with legacy store and is moving to a separate module
called `keycloak-model-legacy`. This impacts number of areas, especially those related to user
federation and custom user providers.

Furthermore, APIs have been consolidated so that the details of storage
layer will be transparent to REST service layer. Specifically, the
services will not be able to differentiate cached and non-cached objects, nor
specifically access federated vs. local storage.

Hence custom extensions that access objects in local storage or cache via `KeycloakSession`
methods must be reviewed. See link:{upgradingguide_link}[{upgradingguide_name}] for details.
```